File: ParallellProgrammering/bird.java
//  One producer, multiple consumers program with monitors
//
// Usage:
//         javac bird.java
//         java numBabyBirds numWorms
class Dish{
	private int value;
	private boolean empty = true;
	private int id;

	public synchronized void take(int id){//invoked by one thread at a time
		this.id = id;
		empty = (value == 0)? true : false;
		notifyAll();//signal all threads waiting on condition empty

		while(empty){
			try{
				wait();//wait on implicit condition empty
			}
			catch(InterruptedException e){
			}
		}
		value--;
		System.out.println("BabyBird "+id+" eats and sleeps");
	}

	public synchronized void put(int value){//invoked by one thread at a time
		while (!empty){
			try{
				wait();//wait on implicit condition !empty
			}
			catch (InterruptedException e){
			}
		}
		System.out.println("Baby "+id+" calls parent");
		this.value = value;
		System.out.println("Parent fills");
		empty = false;
		notifyAll();//signal all threads waiting on condition !empty

	}
}




class BabyBirds extends Thread{
	private Dish dish;
	private int id;
	public BabyBirds(Dish dish, int id){
		this.dish = dish;
		this.id = id;
	}

	public void run(){
		while(true){
			dish.take(id);
			try{
				sleep(1000);
				System.out.println("Baby "+id+" wakes up");
			}
			catch(InterruptedException e){
			}
		}
	}
}



class ParentBird extends Thread{
	private Dish dish;
	private int value;
	public ParentBird (Dish dish, int value){
		this.dish = dish;
		this.value = value;
	}

	public void run(){
		while(true){
			dish.put(value);
		}
	}
}



class bird{// main program -- N consumers and one producer
	private static int MAXBABYBIRDS = 10;
	private static int MAXWORMS = 10;

	public static void main(String args[]){
		Dish dish = new Dish();

		int numBabyBirds = (args.length > 0)? Integer.parseInt(args[0]) : MAXBABYBIRDS;
		int numWorms = (args.length > 1)? Integer.parseInt(args[1]) : MAXWORMS;
		if (numBabyBirds > MAXBABYBIRDS) numBabyBirds = MAXBABYBIRDS;
		if (numWorms > MAXWORMS) numWorms = MAXWORMS;

		ParentBird pb = new ParentBird(dish, numWorms);
		pb.start();
		for(int i = 0; i < numBabyBirds; i++){
			BabyBirds bb = new BabyBirds(dish, i);
			bb.start();
		}
	}
}

File: ParallellProgrammering/bird.c
/* a one producer - multiple consumer using semaphores and threads

   usage on Solaris:
     gcc bird.c -lpthread -lposix4
     a.out numBabyBirds numWorms

*/
#ifndef _REENTRANT
#define _REENTRANT
#endif
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <stdbool.h>
#include <unistd.h>//needed for sleep()

#define SHARED 1
#define MAXBABYBIRDS 10
#define MAXWORMS 10


void *ParentBird(void *);  /* the two threads */
void *BabyBird(void *);

sem_t empty, full;    /* the global semaphores */
int dish;             /* shared buffer         */
int numWorms;

/* main() -- read command line and create threads, then
             print result when the threads have quit */

int main(int argc, char *argv[]) {
  /* thread ids and attributes */
  long l;
  pthread_t pb;
  pthread_t bb[MAXBABYBIRDS];

  pthread_attr_t attr;
  pthread_attr_init(&attr);
  pthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM);

  int numBabyBirds =(argc > 1)? atoi(argv[1]) : MAXBABYBIRDS;
  numWorms =(argc > 2)? atoi(argv[2]) : MAXWORMS;
  if (numBabyBirds > MAXBABYBIRDS) numBabyBirds = MAXBABYBIRDS;
  if (numWorms > MAXWORMS) numWorms = MAXWORMS;

  sem_init(&empty, SHARED, 1);  /* sem empty = 1 */
  sem_init(&full, SHARED, 0);   /* sem full = 0  */

  printf("main started\n");
  pthread_create(&pb, &attr, ParentBird, NULL);
  for (l = 0; l < numBabyBirds; l++){
	  pthread_create(&bb[l], &attr, BabyBird, (void *) l);
  }

    pthread_join(pb, NULL);
    for (l = 0; l < numBabyBirds; l++){
    	  pthread_join(bb[l], NULL);
  }
}

/* parentBird/producer waits until dish is empty, fills it, unlocks for consumers/babyBirds*/
void *ParentBird(void *arg) {
  while(true) {
    sem_wait(&empty);
    printf("parent fills up dish\n");
    dish = numWorms;
    sem_post(&full);
  }

}

/* babyBirds/consumers waits until dish is full, remove one from dish until dish is empty, then unlock
for producer/parentBird*/
void *BabyBird(void *arg) {
   long myid = (long) arg;
  while(true) {
	  sem_wait(&full);
	  if (dish == 0){
		  printf("dish empty, baby %d calls parent\n", myid);
		  sem_post(&empty);
	  }

	  else{
		  printf("baby %d eats a worm and sleeps\n", myid);
		  dish--;
		  sem_post(&full);
		  sleep(1);
		  printf("baby %d is awake\n", myid);
	  }
  }
}
